### PR TITLE
Check possible undefined on LHS IndexedType for logical OR short-circuit

### LINKED ISSUE #29642: Object is possibly 'undefined'
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:** Object undefined

**Code** turn `strictNullChecks` on

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
type EventType = 'click' | 'dblclick'

const handlerMap: { [P in EventType]?: any[] } = {}

function addHandler<P extends EventType>(evType: P) {
  const handlerList = handlerMap[evType] || []
  handlerList.push({}) // Error here: Object is possibly 'undefined'
  handlerMap[evType] = handlerList
}
```

**Expected behavior:** No Error

**Actual behavior:** Shows Error

**Playground Link:**  http://www.typescriptlang.org/play/#src=type%20EventType%20%3D%20'click'%20%7C%20'dblclick'%0Aconst%20handlerMap%3A%20%7B%20%5BP%20in%20EventType%5D%3F%3A%20any%5B%5D%20%7D%20%3D%20%7B%7D%0Afunction%20addHandler%3CP%20extends%20EventType%3E(evType%3A%20P)%20%7B%0A%20%20const%20handlerList%20%3D%20handlerMap%5BevType%5D%20%7C%7C%20%5B%5D%0A%20%20handlerList.push(%7B%7D)%20%2F%2F%20Error%20here%3A%20Object%20is%20possibly%20'undefined'%0A%7D%0A

**Related Issues:**

### PR DESCRIPTION
**Pull request by collin5**
Fixes #29642

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -232,6 +232,18 @@ namespace ts {
         return undefined;
     }
 
+    /**
+     * Returns true if any of the items in array satisfies predicate 'if provided' or is truthy
+     */
+    export function any<T>(array: ReadonlyArray<T>, predicate?: (i: any) => boolean): boolean {
```
Comment (RyanCavanaugh) [inline src/compiler/core.ts]: This function is built-in; it's `Array#some`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -22763,6 +22763,10 @@ namespace ts {
                         getUnionType([extractDefinitelyFalsyTypes(strictNullChecks ? leftType : getBaseTypeOfLiteralType(rightType)), rightType]) :
                         leftType;
                 case SyntaxKind.BarBarToken:
+                    leftType = leftType.flags & TypeFlags.IndexedAccess ?
+                        any(getPropertiesOfType((<IndexedAccessType>leftType).objectType), prop => !!(prop.flags & SymbolFlags.Optional)) ?
```
Comment (weswigham) [inline src/compiler/checker.ts]: I don't think this is right - checking is `any` prop in the indexed access's type is optional and then returning `never` for its type if so seems very wrong - the potential type of the LHS is completely lost here, and it loses the nuance of the individual property accessed. This needs #29317 to be correctly solved - the LHS type should be `leftType & not 0 & not null & not undefined & not ""`, which handles generics appropriately.

Comment (weswigham) [reply src/compiler/checker.ts]: The fact that we don't have more tests this change, as-is, breaks is actually a little disturbing.

**Comment by sandersn** [comment]
I'm cleaning up our PR backlog and this one is pretty stale, so I'm going to close it.

The issue this fixes, #7719, is closed as fixed by #10357, so I'm not really sure if it's *actually* fixed. @collin5 you can consider creating a new PR if it's not. @weswigham, you said this probably needs negated types (from #29317) for a real fix, though. Do you still think that?

**Pull request closed without merging**